adapter/kubernetes/cache.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/runtime"
+	"k8s.io/client-go/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/mixer/pkg/adapter"
+)
+
+type (
+	// internal interface used to support testing
+	cacheController interface {
+		Run(<-chan struct{})
+		GetPod(string) (*v1.Pod, error)
+		HasSynced() bool
+	}
+
+	controllerImpl struct {
+		clientset     kubernetes.Interface
+		logger        adapter.Logger
+		pods          cache.SharedInformer
+		mutationsChan chan resourceMutation
+	}
+
+	// used to send updates to the logger
+	resourceMutation struct {
+		kind eventType
+		obj  interface{}
+	}
+
+	eventType int
+)
+
+const (
+	addition eventType = iota
+	update
+	deletion
+)
+
+// mutationBufferSize sets the limit on how many mutation events can be
+// outstanding at any moment. 100 is chosen as a reasonable default to start.
+// TODO: make this configurable
+const mutationBufferSize = 100
+
+// errorDelay controls how long the logger waits when encountering an error
+// during logging. This should only ever happen when the underlying cache has
+// not yet synced (at which point we need to wait before doing any further
+// processing).
+// TODO: make this configurable
+const errorDelay = 1 * time.Second
+
+// Responsible for setting up the cacheController, based on the supplied client.
+// It configures the index informer to list/watch pods and send update events
+// to a mutations channel for processing (in this case, logging).
+func newCacheController(clientset *kubernetes.Clientset, namespace string, refreshDuration time.Duration, logger adapter.Logger) cacheController {
+	c := &controllerImpl{
+		clientset:     clientset,
+		logger:        logger,
+		mutationsChan: make(chan resourceMutation, mutationBufferSize),
+	}
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts v1.ListOptions) (runtime.Object, error) {
```

Comment:
Yes. It is hard to parse out of the PR description, but in my manual testing with namespace set to "", you can see event updates across namespaces.